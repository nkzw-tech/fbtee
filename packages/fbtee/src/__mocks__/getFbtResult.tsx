import type FbtResult from '../FbtResult.tsx';
import type { NestedFbtContentItems } from '../Types.d.ts';

export default function getFbtResult(
  contents: NestedFbtContentItems,
): FbtResult {
  return (contents?.length === 1 && typeof contents[0] === 'string'
    ? contents[0]
    : contents) as unknown as FbtResult;
}
