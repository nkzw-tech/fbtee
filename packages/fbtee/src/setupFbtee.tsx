import { PatternHash } from '@nkzw/babel-plugin-fbtee';
import FbtResult from './FbtResult.tsx';
import FbtTranslations, { TranslationDictionary } from './FbtTranslations.tsx';
import getFbsResult from './getFbsResult.tsx';
import Hook, { Hooks } from './Hooks.tsx';
import type { IFbtErrorListener, NestedFbtContentItems } from './Types.js';
import IntlViewerContext from './ViewerContext.tsx';

const hasWindow = typeof window !== 'undefined';

const getFbtResult = (
  contents: NestedFbtContentItems,
  hashKey: PatternHash | null | undefined,
  errorListener: IFbtErrorListener | null,
) => {
  const result = new FbtResult(contents, errorListener, hashKey);
  if (hasWindow) {
    return result;
  }

  const resolvedContents = result.getContents();
  return (resolvedContents?.length === 1 &&
  typeof resolvedContents[0] === 'string'
    ? resolvedContents[0]
    : resolvedContents) as unknown as FbtResult;
};

export default function setupFbtee({
  hooks,
  translations,
}: {
  hooks?: Hooks | null;
  translations: TranslationDictionary;
}) {
  FbtTranslations.registerTranslations(translations);

  if (!hooks) {
    hooks = {};
  }

  if (!hooks.getFbtResult) {
    hooks.getFbtResult = getFbtResult;
  }
  if (!hooks.getFbsResult) {
    hooks.getFbsResult = getFbsResult;
  }
  if (!hooks.getTranslatedInput) {
    hooks.getTranslatedInput = FbtTranslations.getTranslatedInput;
  }
  if (!hooks.getViewerContext) {
    hooks.getViewerContext = () => IntlViewerContext;
  }

  Hook.register(hooks);
}
