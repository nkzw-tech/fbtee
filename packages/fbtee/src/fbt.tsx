/// <reference types="../ReactTypes.d.ts" />

import type { FbtTableKey, PatternString } from '@nkzw/babel-plugin-fbtee';
import invariant from 'invariant';
import type { ReactElement } from 'react';
import FbtResult from './FbtResult.tsx';
import type {
  ParamVariationType,
  ValidPronounUsagesType,
} from './FbtRuntimeTypes.tsx';
import FbtTable from './FbtTable.tsx';
import FbtTableAccessor, { FbtTableArg } from './FbtTableAccessor.tsx';
import GenderConst from './GenderConst.tsx';
import getAllSubstitutions from './getAllSubstitutions.tsx';
import Hooks, {
  FbtInputOpts,
  FbtRuntimeInput,
  FbtTableArgs,
  ResolverFn,
} from './Hooks.tsx';
import intlNumUtils from './intlNumUtils.tsx';
import {
  getGenderVariations,
  getNumberVariations,
} from './IntlVariationResolver.tsx';
import list from './list.tsx';
import substituteTokens, { Substitutions } from './substituteTokens.tsx';
import type {
  BaseResult,
  FbtConjunction,
  FbtDelimiter,
  NestedFbtContentItems,
} from './Types.ts';

const ParamVariation: ParamVariationType = {
  gender: 1,
  number: 0,
};

const ValidPronounUsages: ValidPronounUsagesType = {
  object: 0,
  possessive: 1,
  reflexive: 2,
  subject: 3,
};

type ValidPronounUsages =
  (typeof ValidPronounUsages)[keyof typeof ValidPronounUsages];

/**
 * Must match implementation from babel-plugin-fbt/src/fbt-nodes/FbtPronounNode.js
 */
const getPronounGenderKey = (
  usage: ValidPronounUsages,
  gender: GenderConst,
) => {
  switch (gender) {
    case GenderConst.NOT_A_PERSON:
      return usage === ValidPronounUsages.object ||
        usage === ValidPronounUsages.reflexive
        ? GenderConst.NOT_A_PERSON
        : GenderConst.UNKNOWN_PLURAL;

    case GenderConst.FEMALE_SINGULAR:
      return GenderConst.FEMALE_SINGULAR;

    case GenderConst.MALE_SINGULAR:
      return GenderConst.MALE_SINGULAR;

    case GenderConst.UNKNOWN_PLURAL:
      return GenderConst.UNKNOWN_PLURAL;

    case GenderConst.UNKNOWN_SINGULAR:
      return usage === ValidPronounUsages.reflexive
        ? GenderConst.NOT_A_PERSON
        : GenderConst.UNKNOWN_PLURAL;
  }
};

export type Variations =
  | [variation: ParamVariationType['number'], value?: number | null]
  | [variation: ParamVariationType['gender'], value: GenderConst];

export function createRuntime<P, T extends BaseResult | string>({
  getResult,
  param,
  plural,
}: {
  getResult: ResolverFn<T>;
  param: (label: string, value: P, variations?: Variations) => FbtTableArg;
  plural: (count: number, label?: string | null, value?: P) => FbtTableArg;
}) {
  const cachedResults = new Map<PatternString, T>();
  return Object.assign(
    (_: string, __?: string, ___?: unknown) => {
      throw new Error(
        `fbt must be used with its corresponding babel plugin. Please install the babel plugin and try again.`,
      );
    },
    {
      _: (
        inputTable: FbtRuntimeInput,
        inputArgs?: FbtTableArgs | null,
        options?: FbtInputOpts | null,
      ): T => {
        let { args, table } = Hooks.getTranslatedInput({
          args: inputArgs || null,
          options: options || null,
          table: inputTable,
        });

        let substitutions: Substitutions | null = null;

        if (table.__vcg != null) {
          if (!args) {
            args = [];
          }
          args.unshift(
            FbtTableAccessor.getGenderResult(
              getGenderVariations(Hooks.getViewerContext().GENDER),
              null,
            ),
          );
        }

        const tokens: Array<FbtTableKey> = [];
        if (args) {
          if (typeof table !== 'string') {
            const newPattern = FbtTable.access(table, args, 0, tokens);
            if (newPattern) {
              table = newPattern;
            }
          }

          invariant(table !== null, 'Table access failed');
          substitutions = getAllSubstitutions(args);
        }

        const patternString = Array.isArray(table) ? table[0] : table;
        if (typeof patternString !== 'string') {
          throw new Error(
            `Table access did not result in string: ${table === undefined ? 'undefined' : JSON.stringify(table)}, Type: ${typeof table}`,
          );
        }

        const cachedFbt = cachedResults.get(patternString);
        if (cachedFbt && !substitutions) {
          return cachedFbt;
        } else {
          const fbtContent = substituteTokens(patternString, substitutions);
          const result = getResult(
            typeof fbtContent === 'string'
              ? [fbtContent]
              : (fbtContent as NestedFbtContentItems),
            options?.hk,
            Hooks.getErrorListener({
              hash: options?.hk,
              translation: patternString,
            }),
          );
          if (!substitutions) {
            cachedResults.set(patternString, result);
          }
          return result;
        }
      },
      _enum: (
        value: FbtTableKey,
        range: {
          [enumKey: string]: string;
        },
      ) => {
        if (process.env.NODE_ENV === 'development') {
          invariant(value in range, 'invalid value: %s', value);
        }
        return FbtTableAccessor.getEnumResult(value);
      },
      _implicitParam: (label: string, value: P, variations?: Variations) =>
        param(label, value, variations),
      _list: (
        label: string,
        items: ReadonlyArray<string | ReactElement | null | undefined>,
        conjunction?: FbtConjunction,
        delimiter?: FbtDelimiter,
      ) => [
        null,
        {
          [label]: list(items, conjunction, delimiter),
        },
      ],

      _name: (label: string, value: P, gender: GenderConst) =>
        FbtTableAccessor.getGenderResult(getGenderVariations(gender), {
          [label]: value,
        }),

      _param: param,
      _plural: plural,

      _pronoun: (
        usage: ValidPronounUsages,
        gender: GenderConst,
        options?: {
          human?: 1;
        } | null,
      ) => {
        invariant(
          gender !== GenderConst.NOT_A_PERSON || !options || !options.human,
          'Gender cannot be GenderConst.NOT_A_PERSON if you set "human" to true',
        );
        return FbtTableAccessor.getPronounResult(
          getPronounGenderKey(usage, gender),
        );
      },

      _subject: (value: GenderConst) =>
        FbtTableAccessor.getGenderResult(getGenderVariations(value), null),
    } as const,
  );
}

export default createRuntime<string | number, FbtResult>({
  getResult: Hooks.getFbtResult,
  param: (label: string, value: number | string, variations?: Variations) => {
    const substitution = { [label]: value };
    if (variations) {
      if (variations[0] === ParamVariation.number) {
        const number = variations.length > 1 ? variations[1] : value;
        invariant(typeof number === 'number', 'fbt.param expected number');

        const variation = getNumberVariations(number); // this will throw if `number` is invalid
        if (typeof value === 'number') {
          substitution[label] =
            intlNumUtils.formatNumberWithThousandDelimiters(value);
        }
        return FbtTableAccessor.getNumberResult(variation, substitution);
      } else if (variations[0] === ParamVariation.gender) {
        const gender = variations[1];
        invariant(gender != null, 'expected gender value');
        return FbtTableAccessor.getGenderResult(
          getGenderVariations(gender),
          substitution,
        );
      } else {
        invariant(false, 'Unknown invariant mask');
      }
    } else {
      return FbtTableAccessor.getSubstitution(substitution);
    }
  },
  plural: (count: number, label?: string | null, value?: number | string) =>
    FbtTableAccessor.getNumberResult(
      getNumberVariations(count),
      label
        ? {
            [label]:
              typeof value === 'number'
                ? intlNumUtils.formatNumberWithThousandDelimiters(value)
                : value ||
                  intlNumUtils.formatNumberWithThousandDelimiters(count),
          }
        : null,
    ),
});
