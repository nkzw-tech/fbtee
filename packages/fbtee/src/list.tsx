/// <reference types="../ReactTypes.d.ts" />

import type { ReactElement, ReactNode } from 'react';
import fbt from './fbt.tsx';
import type { FbtConjunction, FbtDelimiter } from './Types.ts';

export default function list(
  items: ReadonlyArray<string | ReactElement | null | undefined>,
  conjunction: FbtConjunction = 'and',
  delimiter: FbtDelimiter = 'comma',
): ReactNode {
  // Ensure the local version of `fbt` is used instead of auto-importing `fbtee`.
  // eslint-disable-next-line no-unused-expressions, @typescript-eslint/no-unused-expressions
  fbt;

  items = items.filter(Boolean);

  const count = items.length;
  if (count === 0) {
    return '';
  } else if (count === 1) {
    return items[0];
  }

  const lastItem = items.at(-1);
  let output: ReactNode = items[0];

  for (let index = 1; index < count - 1; index++) {
    switch (delimiter) {
      case 'semicolon':
        output = (
          <fbt desc='A list of items of various types, for example: "San Francisco; London; Tokyo". {previous items} and {following items} are themselves lists that contain one or more items.'>
            <fbt:param name="previous items">{output}</fbt:param>
            {'; '}
            <fbt:param name="following items">{items[index]}</fbt:param>
          </fbt>
        );
        break;
      case 'bullet':
        output = (
          <fbt desc='A list of items of various types separated by bullets, for example: "San Francisco \u2022 London \u2022 Tokyo". {previous items} and {following items} are themselves lists that contain one or more items.'>
            <fbt:param name="previous items">{output}</fbt:param> &bull;{' '}
            <fbt:param name="following items">{items[index]}</fbt:param>
          </fbt>
        );
        break;
      default:
        output = (
          <fbt desc='A list of items of various types separated by commas, for example: "San Francisco, London, Tokyo". {previous items} and {following items} are themselves lists that contain one or more items.'>
            <fbt:param name="previous items">{output}</fbt:param>
            {', '}
            <fbt:param name="following items">{items[index]}</fbt:param>
          </fbt>
        );
    }
  }

  switch (conjunction) {
    case 'and':
      return (
        <fbt desc='A list of items of various types, for example: "item1, item2, item3 and item4"'>
          <fbt:param name="list of items">{output}</fbt:param>
          and
          <fbt:param name="last item">{lastItem}</fbt:param>
        </fbt>
      );

    case 'or':
      return (
        <fbt desc='A list of items of various types, for example: "item1, item2, item3 or item4"'>
          <fbt:param name="list of items">{output}</fbt:param>
          or
          <fbt:param name="last item">{lastItem}</fbt:param>
        </fbt>
      );

    case 'none':
      switch (delimiter) {
        case 'semicolon':
          return (
            <fbt desc='A list of items of various types, for example: "San Francisco; London; Tokyo". {previous items} itself contains one or more items.'>
              <fbt:param name="previous items">{output}</fbt:param>
              {'; '}
              <fbt:param name="last item">{lastItem}</fbt:param>
            </fbt>
          );
        case 'bullet':
          return (
            <fbt desc='A list of items of various types separated by bullets, for example: "San Francisco \u2022 London \u2022 Tokyo". {previous items} contains one or more items.'>
              <fbt:param name="list of items">{output}</fbt:param> &bull;{' '}
              <fbt:param name="last item">{lastItem}</fbt:param>
            </fbt>
          );
        default:
          return (
            <fbt desc='A list of items of various types, for example: "item1, item2, item3, item4"'>
              <fbt:param name="list of items">{output}</fbt:param>
              {', '}
              <fbt:param name="last item">{lastItem}</fbt:param>
            </fbt>
          );
      }
    default: {
      conjunction satisfies never;
      throw new Error(
        `Invalid conjunction ${conjunction} provided to '<fbt:list>'.`,
      );
    }
  }
}

export function List({
  conjunction,
  delimiter,
  items,
}: {
  conjunction?: FbtConjunction;
  delimiter?: FbtDelimiter;
  items: Array<string | React.ReactElement | null | undefined>;
}) {
  return list(items, conjunction, delimiter);
}
