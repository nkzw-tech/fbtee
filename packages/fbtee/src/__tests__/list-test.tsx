import { describe, expect, it } from '@jest/globals';
import getFbtResult from '../__mocks__/getFbtResult.tsx';
import fbt from '../fbt.tsx';
import list from '../list.tsx';
import setupFbtee from '../setupFbtee.tsx';
import IntlViewerContext from '../ViewerContext.tsx';

setupFbtee({
  hooks: {
    getFbtResult,
    getViewerContext: () => IntlViewerContext,
  },
  translations: { en_US: {} },
});

describe('list', () => {
  it('should handle an empty list', () => {
    expect(list([])).toBe('');
  });
  it('should handle a list full of null/undefined items', () => {
    expect(list([null, undefined])).toBe('');
  });
  it('should handle a single item', () => {
    expect(list(['first'])).toBe('first');
  });
  it('should handle two items', () => {
    expect(list(['first', 'second'])).toBe('first and second');
  });
  it('should handle three items', () => {
    expect(list(['first', 'second', 'third'])).toBe('first, second and third');
  });
  it('should handle a bunch of items', () => {
    expect(list(['1', '2', '3', '4', '5', '6', '7', '8'])).toBe(
      '1, 2, 3, 4, 5, 6, 7 and 8',
    );
  });
  it('should handle a bunch of items, some of which are null/undefined', () => {
    expect(
      list(['1', '2', '3', '4', null, '5', undefined, '6', '7', '8']),
    ).toBe('1, 2, 3, 4, 5, 6, 7 and 8');
  });
  it('should handle no conjunction', () => {
    expect(list(['first', 'second', 'third'], 'none')).toBe(
      'first, second, third',
    );
  });
  it('should handle optional delimiter', () => {
    expect(list(['first', 'second', 'third'], 'none', 'semicolon')).toBe(
      'first; second; third',
    );
  });
  it('should handle bullet delimiters', () => {
    expect(list(['first', 'second', 'third'], 'none', 'bullet')).toBe(
      'first \u2022 second \u2022 third',
    );
  });
});

test('fbt.list()', () => {
  expect(
    fbt(
      'Available Locations: ' +
        // @ts-expect-error
        fbt.list('locations', ['Tokyo', 'London', 'Vienna']),
      'Lists',
    ),
  ).toMatchSnapshot();
});

test('<fbt:list>', () => {
  // eslint-disable-next-line no-unused-expressions, @typescript-eslint/no-unused-expressions
  fbt;

  expect(
    <fbt desc="Lists">
      Available Locations:{' '}
      <fbt:list items={['Tokyo', 'London', 'Vienna']} name="locations" />.
    </fbt>,
  ).toMatchSnapshot();
});
