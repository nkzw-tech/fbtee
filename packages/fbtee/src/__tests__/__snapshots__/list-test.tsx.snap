// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<fbt:list> 1`] = `"Available Locations: Tokyo, London and Vienna."`;

exports[`fbt.list() 1`] = `"Available Locations: Tokyo, London and Vienna"`;
